[21.05] Pl 130693 ceph luminous

osnyx opened the pull request:
@flyingcircusio/release-managers

## Release process

Impact:

none

Changelog:

* Provide the Ceph Luminuous update using a feature-flag mechanism. (PL-130693)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

This is a comprehensive update, however, it doesn't touch too many things from an architectural perspective. We know about some of the deficits of our current implementation (like daemons running as root, but on dedicated machines) that did not change too much.

One aspect of attack surface is covered: the manager HTTP server is bound to localhost instead of the default to listen on all addresses.

Otherwise our key management is a bit simplistic, to allow for smooth automated management and also as things run as root anyway separating keys further seems overkill.

- [x] Security requirements tested? (EVIDENCE)

The whole dev cluster has been updated to this PR's state, individual hosts and Ceph components have been updated to the new version step by step. This is true for both the Ceph storage servers as well as the Ceph client components part of other infrastructure components (kvm, backup).
Major regular maintenance and tasks and operations steps have been replicated on the updated cluster, documentation has been adapted wherever necessary. A stress test (chaosmonkey) has been launched on the dev cluster with no apparent failures.
Several disk format changes, rebuilds and recoverys went through successfully without data loss.
Performance did even increase with the use of Bluestore (beneficial to availability).
Automated NixOS tests

ctheune commented on pkgs/overlay.nix:
```diff
@@ -35,10 +35,20 @@ in {
       boost = super.boost155;
   });
+  ceph-luminous = (super.callPackage ./ceph/luminous {
+    # TODO: try newer boost versions
```
not needed. remove todo marker.

ctheune commented on pkgs/overlay.nix:
```diff
@@ -414,6 +424,43 @@ in {
+        # work around a weird pythonImportsCheck failure
+        #buildInputs = oldattrs.propagatedBuildInputs;
+        #pythonImportsCheck = [];
+        pythonImportsCheckPhase = ''
```
I think this can be simplified.

ctheune commented on pkgs/fc/ceph/default.nix:
```diff
@@ -11,15 +11,15 @@ py.buildPythonApplication rec {
     python3Packages.requests
   ];
 
+  # FIXME: only temporary
```
well, fix it then ;)

ctheune commented on pkgs/fc/ceph/src/fc/ceph/main.py:
```diff
@@ -1,12 +1,19 @@
+from fc.ceph import Environment
+
+CONFIG_FILE_PATH = Path("/etc/ceph/fc-ceph.conf")
+# FIXME: move to OSD module exported constants
```
please check

ctheune commented on pkgs/fc/ceph/src/fc/ceph/main.py:
```diff
@@ -92,8 +119,17 @@ def main(args=sys.argv[1:]):
+        "--target-objectstore-type",
+        "-T",
+        choices=OBJECTSTORE_TYPES,
+        # FIXME: do I explicitly need to add the choices to the help text?
```
check

ctheune commented on pkgs/fc/ceph/src/fc/ceph/subsystems.py:
```diff
@@ -0,0 +1,12 @@
+class VersionedSubsystem(object):
```
this seems to  be a duplicate of fc.ceph.VersionedSubsystems. I think this file can just go away.

ctheune commented on nixos/roles/backyserver.nix:
```diff
@@ -8,6 +8,7 @@ let
   enc = config.flyingcircus.enc;
 
   backy = pkgs.callPackage ../../pkgs/backy.nix { };
+  ceph_common = import ./ceph/ceph-common.nix { inherit lib pkgs; };
```
please refactor to leverage fclib.ceph instead of introducing a new pattern

ctheune commented on nixos/roles/ceph/osd.nix:
```diff
@@ -66,6 +71,10 @@ in
           osd op history size = 1000
           osd op history duration = 43200
 
+          # we currently have (too) many PGs in our pools, make sure the clsuter still
```
typo

ctheune commented on tests/default.nix:
```diff
@@ -48,6 +50,10 @@ in {
   kibana7 = callTest ./kibana.nix { version = "7"; };
   k3s = callTest ./k3s {};
   kvm_host = callTest ./kvm_host.nix {};
```
Suggestin: rename to kvm_host_jewel-ceph-server-jewel and avoid a version-less alias

osnyx committed:
ignore visual studio code files

osnyx committed:
ceph: add luminous package

- based on old upstream nixpkgs packaging
- still built against python2 as the new ceph-mgr does not support
  python3 yet
- requires vendoring of a lot of python27 packages to backport them to the
  last version support python-2.7:

    - cheroot
    - cherrypy
    - jaraco_text
    - pecan
    - portend
    - pypytools
    - pyquery
    - tempora
    - waitress
    - webtest
    - zc_lockfile

    Additionally, tests for the following python2 packages needed to be
    disabled:

    - routes
    - webtest
    - zc_lockfile
    - cherrypy

    zc_lockfile has issues with test case discovery, all other cases of
    disabling tests are due to failing imports of modules containing a "."
    in their name.

- python27Packages.Cython: 0.29.22 -> 0.29.28
  this matches the Cython version used for generating the release *.pyx
  files of ceph-12.2.13

osnyx committed:
ceph: adapt fc.ceph to support luminous (and also jewel in parallel)

Summary of changes:

- refactor and split up subsystems to serve different ceph versions:
    - restructure `manage` subsystems into mon, mgr (Luminous only), osd
    - introduce subsystem versioning and dispatching logic (Environment,
      VersionedSubsystem)
    - don't rely on used executables to be in system PATH anymore,
      instead read path and module version from /etc/ceph/fc-ceph.conf
      provided by platform
    - introduce new mgr component with own key
- distinguish between OSD types bluestore and filestore
    - type-specific `create` subcommands
    - OSD rebuilding supports changing between objectstore types
- fc-ceph/osd: simplify destruction by using new purge command
- fc-ceph/osd disk layout with WAL backup

    Bluestore OSDs with external WALs now store a backup of that WAL on
    their main disk when deactivating, and restore it from there when
    activating again. This enables migration of disks between hosts.

    A backwards-incompatible disk layout change was required.

    Part of PL-130693 and PL-130677
- fc-ceph/maintenance: adapt maintenance checking to differing json
  status format
- fc-ceph: Show usage information on incomplete subcommands

    When providing a subcommand, but no action to it, an overview of all
    actions of this subcommand is shown.
    Includes tests.

    Inspired by https://github.com/flyingcircusio/batou/pull/309

osnyx committed:
ceph: adapt platform to support luminous (and jewel in parallel)

summary:
- ceph: make ceph release of roles, client, and server configurable

    Each individual ceph component (ceph_mon, ceph_osd, ceph_rgw) can be
    parameterised via it's role's new
    `cephRelease` parameter to be either of the jewel or luminous release.
    This parameter is also passed to the `flyingcircus.services.ceph`
    services, which gained a new `package` option for that purpose.

    Due to the ceph package being added to the system path, once at least a
    single active role is on luminous, the system PATH package will be set
    to luminous as well.

    The default release remains "jewel" for now, enabling the
    unparameterised ceph jewel NixOS test to pass.

    The same is true for the roles using ceph as a client (kvm_host,
    backyserver).
- the definition of the global ceph package in path on
  `flyingcircus.service.ceph.server` machines uses a custom
  type, enabling each ceph component role to define that option separately
  and then merging these definitions in a sensible way.
- fclib gets a `fclib.ceph` module with common constants, helper
  functions, types, and release-name <-> package mapping
- provide tooling path and subsystem ceph version to fc-ceph via
  /etc/ceph/fc-ceph.conf
  `fc-ceph key generate-client-keyring` is part of the activation script
  on all ceph clients as well. As a result, at least an fc-ceph default
  path needs to exist on all clients.
- alias `ceph` to `ceph-luminous` during migration period
- update and adapt fc.qemu
- disable automatic crush map reorganisation from messing up pools

  As a new feature, ceph luminous can detect the device class (ssd, hdd,
  nvme) of a device automatically and then move devices of different
  classes into the same host pool, as they can then be differentiated by
  their new class attribute.
  This messes with our current shadow root bucket structure and thus needs
  to be disabled at least during the upgrade.

  Device class detection is kept enabled as it does not change the bucket
  structure by itself in a harmful way.
- fc.agent/createvm: Make output more stable to ceph version changes

  `fc-create-vm` invokes the `rbd` command directly and makes use of its
  output, both for parsing data as well as for displaying it. The output
  of `rbd snap ls` differs depending on the ceph version used, making both
  parsing as well as matching of data (in tests) unreliable.
  This commit changes a single problematic command to make use of the
  JSON-formatted output to reliably extract and display the same data
  independent of ceph version.
  This re-uses the JSONRunner tooling created for fc-ceph.
- ceph/osd: increase hard limit on PGs per OSD
- ceph/mgr: enable dashboard on localhost
- fc.util-physical: make `rbd-locktool` use correct rbd tool
  regression: only when `rbd-locktool` is invoked via fc-ceph, otherwise
  the in-path command still always uses the default `ceph` package. This
  is known to cause problems in mixed clusters.
- fc.check-ceph: split up in ceph-release specific code bases and adapt
  to changed Luminous status format.
  The attributes "detail", "summary", and "overall_status" have vanished
  from the json output of `ceph status` and needed to be reconstructed
  from other values.

osnyx committed:
ceph: test coverage for ceph luminous support

summary:
- split and adapt tests targeting either ceph jewel or luminous
    - luminous test cover both filestore and bluestore, internal and
      external journals, new mgr component, rebuilding and recreating
      OSDs
    - kvm_host test can be parameterised to either use the old jewel or
      the new luminous client tooling on top of a luminous cluster
- tests/ceph-luminous: update assert_clean_cluster

  Adjust to luminous style.
  Move to JSON output parsing of `ceph -s`.
  Include checking for mgrs as well.
- tests/kvm_host_ceph_luminous: don't fail if all tests are deselected

    This is relevant when providing a test case selection argstring, e.g.
    `--arg useCheckout true --argstr testOpts "-k test_simple_vm_lifecycle_start_stop" tests/kvm_host_ceph-luminous.nix`
    like suggested in fc.qemu/HACKING.txt

osnyx committed:
physical hosts: in-PATH rbd-locktool uses correct ceph

By inheriting the ceph package (and release) from the client ceph
service, rbd-locktool can avoid incompatibilities when running on a
(partially) newer cluster.

osnyx committed:
ceph/mon: drop pg max object skew warning

The warning check `mon pg warn max object skew` has been dropped by
upstream in Ceph Luminous. As the reason for the removal isn't an
improvement in object assignment making skew disappear, but problems
with the check and its quality of implementation (see
https://tracker.ceph.com/issues/20730), it is okay to already remove it
for Ceph Jewel hosts as well.

dpausp commented on nixos/roles/backyserver.nix:
```diff
@@ -22,6 +22,10 @@ let
 
   restoreSingleFiles = pkgs.callPackage ../../pkgs/restore-single-files {};
 
+  backy_rbd_versioned = cephReleaseName: {
```
style: indentation, camelCase

dpausp commented on nixos/roles/ceph/osd.nix:
```diff
@@ -120,19 +128,41 @@ in
+      in {
+        # fc-ceph OSD
+        OSDManager = osdSettings;
+        # FIXME: the module uses the `rbd` binary. While it'd be safer to handle it's
```
remove FIXME
